storm-core/src/jvm/org/apache/storm/daemon/supervisor/Slot.java

```diff
@@ -0,0 +1,769 @@
+        }
+        //Both assignments are null, just wait
+        if (dynamicState.profileActions != null && !dynamicState.profileActions.isEmpty()) {
+            //Nothing is scheduled here so throw away all of the profileActions
+            LOG.warn("Dropping {} no topology is running", dynamicState.profileActions);
+            dynamicState = dynamicState.withProfileActions(Collections.<TopoProfileAction> emptySet(), Collections.<TopoProfileAction> emptySet());
+        }
+        Time.sleep(1000);
+        return dynamicState;
+    }
+    
+    private final AtomicReference<LocalAssignment> newAssignment = new AtomicReference<>();
+    private final AtomicReference<Set<TopoProfileAction>> profiling =
+            new AtomicReference<Set<TopoProfileAction>>(new HashSet<TopoProfileAction>());
+    private final StaticState staticState;
+    private final IStormClusterState clusterState;
+    private volatile boolean done = false;
+    private volatile DynamicState dynamicState;
+    private final AtomicReference<Map<Long, LocalAssignment>> cachedCurrentAssignmants;
+    
+    public Slot(ILocalizer localizer, Map<String, Object> conf, 
+            ContainerLauncher containerLauncher, String host,
+            int port, LocalState localState,
+            IStormClusterState clusterState,
+            ISupervisor iSupervisor,
+            AtomicReference<Map<Long, LocalAssignment>> cachedCurrentAssignmants) throws Exception {
+        super("SLOT_"+port);
+
+        this.cachedCurrentAssignmants = cachedCurrentAssignmants;
+        this.clusterState = clusterState;
+        Map<Integer, LocalAssignment> assignments = localState.getLocalAssignmentsMap();
+        LocalAssignment currentAssignment = null;
+        if (assignments != null) {
+            currentAssignment = assignments.get(port);
+        }
+        Container container = null;
+        if (currentAssignment != null) { 
+            container = containerLauncher.recoverContainer(port, currentAssignment, localState);
+        }
+        
+        LocalAssignment newAssignment = currentAssignment;
+        if (currentAssignment != null && container == null) {
+            currentAssignment = null;
+            //Assigned something but it is not running
+        }
+        
+        dynamicState = new DynamicState(currentAssignment, container, newAssignment);
+        staticState = new StaticState(localizer, 
+                Utils.getInt(conf.get(Config.SUPERVISOR_WORKER_TIMEOUT_SECS)) * 1000,
+                Utils.getInt(conf.get(Config.SUPERVISOR_WORKER_START_TIMEOUT_SECS)) * 1000,
+                Utils.getInt(conf.get(Config.SUPERVISOR_WORKER_SHUTDOWN_SLEEP_SECS)) * 1000,
+                Utils.getInt(conf.get(Config.SUPERVISOR_MONITOR_FREQUENCY_SECS)) * 1000,
+                containerLauncher,
+                host,
+                port,
+                iSupervisor,
+                localState);
+        this.newAssignment.set(dynamicState.newAssignment);
```

Comment:
We could, but that is meant to be a user facing API, and so I don't see a lot of value in trying to maintain the abstraction here.